README.rst
===============================
dgim
===============================

.. image:: https://badge.fury.io/py/dgim.png
    :target: http://badge.fury.io/py/dgim

.. image:: https://travis-ci.org/simondolle/dgim.png?branch=master
        :target: https://travis-ci.org/simondolle/dgim

.. image:: https://pypip.in/d/dgim/badge.png
        :target: https://pypi.python.org/pypi/dgim


Python implementation of the DGIM algorithm: 
a compact datastructure to estimate the number of *True* statements in the last N elements of a boolean stream.

Features
--------

* Estimation of the number of "True" statements in the last N element of a boolean stream
* Low memory footprint.
* Tunable error rate (the lower the error rate, the higher the memory footprint)

Applications
------------

When processing large streams of data such as clicks streams, server logs, financial streams.
It is often necessary to maintain statistics about the N latest elements.
If N is big or if you have many streams to process, it is not possible to store
the N latest elements.

In such situations, if the processed stream is made of boolean,
the DGIM algorithm can help you estimate the number of *True* statements
in the last elements.

For instance, if the stream is made of server logs, 
DGIM algorithm can estimate the proportion of visits that come from search engines.
(as opposed to direct access, or access through paid search)


Installation
------------

At the command line::

    $ pip install dgim

Usage
-----

Sample code::

  from dgim import Dgim
  dgim = Dgim(N=32, error_rate=0.1)
  for i in range(100):
      dgim.update(True)
  dgim_result = dgim.get_count() # 30 (exact result is 32)


Documentation
-------------

https://dgim.readthedocs.org.


License
-------

The project is licensed under the BSD license.

Authors
-------

* Simon Dollé <simon.dolle@gmail.com>
* No other contributor yet. Why not joining?

How to contribute
-----------------

#. Check for open issues or open a fresh issue to start a discussion around a feature idea or a bug.
#. Fork `the repository`_ on GitHub to start making your changes to the **master** branch (or branch off of it).
#. Write a test which shows that the bug was fixed or that the feature works as expected.
#. Send a pull request and bug the maintainer until it gets merged and published. :) Make sure to add yourself to AUTHORS_.

References
----------
- Datar, Mayur, et al. "Maintaining stream statistics over sliding windows."
  SIAM Journal on Computing 31.6 (2002): 1794-1813.
- Rajaraman, Anand, and Jeffrey David Ullman. Mining of massive datasets. Cambridge University Press, 2011. Chapter 4. http://infolab.stanford.edu/~ullman/mmds/ch4.pdf
- Mining of Massive Datasets Coursera MOOC: http://infolab.stanford.edu/~ullman/mmds/ch4.pdf

.. _`the repository`: http://github.com/simondolle/dgim
.. _AUTHORS: https://github.com/simondolle/dgim/blob/master/AUTHORS.rst



HISTORY.rst
.. :changelog:

History
-------

0.2.0 (2015-01-05)
---------------------

* Improved documentation
* Make most methods and attribute private.

0.1.0 (2015-01-04)
---------------------

* First release on PyPI.


tests/test_dgim.py
import unittest
import itertools
from collections import deque
from dgim import Dgim


class TestDgim(unittest.TestCase):
    def test_get_count(self):
        dgim = Dgim(12)
        stream = iter([
            False, False, True, False, True, True, True, False, True,
            True, False, False, True, False, True, True, False
        ])
        for elt in stream:
            dgim.update(elt)
        self.assertEquals(7, dgim.get_count())

    def test_get_count_without_update(self):
        """Example from chapter 4 of "Mining of Massing Datasets"""
        crt_timestamp = 65
        # hand built dgim
        queues = [
            deque([crt_timestamp - 1, crt_timestamp - 2]),
            deque([crt_timestamp - 4]),
            deque([crt_timestamp - 8]),
            deque()
        ]

        dgim = Dgim(10)
        dgim._timestamp = crt_timestamp
        dgim._queues = queues
        self.assertEquals(6, dgim.get_count())

    def test_count_empty_stream(self):
        dgim = Dgim(10)
        self.assertEqual(0, dgim.get_count())

    def test_N_is_null(self):
        dgim = Dgim(0)
        stream = iter([True, False, False, True])
        for elt in stream:
            dgim.update(elt)
        self.assertEquals(0, dgim.get_count())

    def test_N_is_one(self):
        dgim = Dgim(1)
        dgim.update(True)
        self.assertEqual(1, dgim.get_count())
        dgim.update(False)
        self.assertEqual(0, dgim.get_count())

    def test_N_is_two(self):
        dgim = Dgim(2)
        dgim.update(True)
        self.assertEqual(1, dgim.get_count())
        dgim.update(True)
        self.assertEqual(2, dgim.get_count())
        dgim.update(True)
        self.assertEqual(2, dgim.get_count())

    def test_bucket_drop(self):
        crt_timestamp = 65
        # hand built dgim
        queues = [
            deque([crt_timestamp - 1, crt_timestamp - 2]),
            deque([crt_timestamp - 4]),
            deque()
        ]
        dgim = Dgim(6)
        dgim._timestamp = crt_timestamp
        dgim._queues = queues
        dgim._oldest_bucket_timestamp = crt_timestamp - 4
        self.assertEquals(3, dgim.nb_buckets)
        dgim.update(0)
        self.assertEquals(3, dgim.nb_buckets)
        dgim.update(0)
        self.assertEquals(2, dgim.nb_buckets)

    def test_only_zeros(self):
        dgim = Dgim(10)
        for elt in itertools.repeat(False, 1000):
            dgim.update(elt)
            self.assertEqual(0, dgim.get_count())

    def test_r_computation(self):
        dgim = Dgim(10, 0.5)
        self.assertEqual(2, dgim._r)

        dgim = Dgim(10, 0.1)
        self.assertEqual(10, dgim._r)

    def test_invalid_error_rates(self):
        self.assertRaises(ValueError, Dgim, 10, 0)
        self.assertRaises(ValueError, Dgim, 10, -0.1)
        self.assertRaises(ValueError, Dgim, 10, 1.1)

    def test_is_bucket_too_old(self):
        dgim = Dgim(10)
        dgim._timestamp = 15
        self.assertFalse(dgim._is_bucket_too_old(6))
        self.assertTrue(dgim._is_bucket_too_old(5))
        self.assertTrue(dgim._is_bucket_too_old(16))

        dgim._timestamp = 5
        self.assertFalse(dgim._is_bucket_too_old(16))
        self.assertTrue(dgim._is_bucket_too_old(15))


tests/test_dgim_quality.py
import unittest
import itertools
from collections import deque

from dgim import Dgim
from dgim.utils import generate_random_stream


class ExactAlgorithm(object):
    """Exact algorithm to count the number of "True"
    in the last N elements of a boolean stream."""
    def __init__(self, N):
        """Constructor
        :param N: size of the sliding window
        :type N: int
        """
        self.N = N
        self.sliding_window = deque()

    def update(self, elt):
        """Update the stream with one element.
        :param elt: the latest element of the stream
        :type elt: bool
        """
        self.sliding_window.append(elt)
        if len(self.sliding_window) > self.N:
            self.sliding_window.popleft()

    def get_count(self):
        """Returns an estimate of the number of "True"
        in the last N elements of the stream.
        :returns: int
        """
        return sum(self.sliding_window)


class TestDgimQuality(unittest.TestCase):

    def check_quality_settings(self, dgim, stream):
        """Compare the result "e" returned by dgim with the exact result
        "c" on a stream which elements are 0 or 1.
        The test fails if the dgim result "e" is not in the expected bounds.
        0.5 * c <= e <= 1.5 * c

        :param dgim: the Dgim instance to test
        :type dgim: Dgim
        :param stream: the stream to use. It should contains only 0 or 1 as elements.
        :type stream: iterator
        """
        exact_algorithm = ExactAlgorithm(dgim.N)
        for elt in stream:
            dgim.update(elt)
            exact_algorithm.update(elt)

            exact_result = exact_algorithm.get_count()
            error = abs(dgim.get_count() - exact_result)
            self.assertTrue(error <= dgim.error_rate * exact_result)

    def test_nominal_case(self):
        dgim = Dgim(N=100, error_rate=0.5)
        stream = generate_random_stream(length=10000)
        self.check_quality_settings(dgim=dgim, stream=stream)

    def test_large_N(self):
        dgim = Dgim(N=10000, error_rate=0.5)
        stream = generate_random_stream(length=2000)
        self.check_quality_settings(dgim=dgim, stream=stream)

    def test_short_stream(self):
        dgim = Dgim(N=1000, error_rate=0.5)
        # stream is shorter than N
        stream = generate_random_stream(length=100)
        self.check_quality_settings(dgim=dgim, stream=stream)

    def test_N_is_one(self):
        dgim = Dgim(N=1, error_rate=0.5)
        stream = generate_random_stream(length=10)
        self.check_quality_settings(dgim=dgim, stream=stream)

    def test_N_is_two(self):
        dgim = Dgim(N=2, error_rate=0.5)
        stream = generate_random_stream(length=100)
        self.check_quality_settings(dgim=dgim, stream=stream)

    def test_low_error_rate_case(self):
        dgim = Dgim(N=100, error_rate=0.01)
        stream = generate_random_stream(length=1000)
        self.check_quality_settings(dgim=dgim, stream=stream)

    def test_only_true_case(self):
        dgim = Dgim(N=100, error_rate=0.5)
        stream = itertools.repeat(True, 10000)
        self.check_quality_settings(dgim=dgim, stream=stream)


bench_runtime.py
import time
import math
import matplotlib.pyplot as plt

from dgim.dgim import Dgim
from dgim.utils import generate_random_stream

def measure_update_time(N, iterations):
    dgim = Dgim(N)
    # initialization
    for elt in generate_random_stream(N):
        dgim.update(elt)
    time_start = time.time()
    bucket_count = 0
    for elt in generate_random_stream(iterations):
        dgim.update(elt)
        bucket_count += dgim.nb_buckets
    time_stop = time.time()
    return time_stop - time_start, bucket_count/float(iterations)

def run_update_benchmark():
    times = []
    bucket_counts = []
    for i in range(24):
        time, bucket_count = measure_update_time(2 ** i, iterations=100000)
        print 2 ** i, time
        times.append((2 ** i, time))
        bucket_counts.append(bucket_count)
    plt.plot([n for n, time in times], [time for n, time in times])
    #plt.plot(bucket_counts, [time for n, time in times])
    plt.show()

if __name__ == "__main__":
    run_update_benchmark()


dgim/dgim.py
import math
from collections import deque

class Dgim(object):
    """An implementation of the DGIM algorithm.
    It estimates the number of "True" present the last N elements
    of a boolean stream.
    The datastructure it uses is very compact and
    has a memory complexity of O(log(N)^2).

    The algorithm is described in:
    Datar, Mayur, et al. "Maintaining stream statistics over sliding windows."
    SIAM Journal on Computing 31.6 (2002): 1794-1813.

    An explanation of the algorithm can also be found here:
    http://infolab.stanford.edu/~ullman/mmds/ch4.pdf
    """

    def __init__(self, N, error_rate=0.5):
        """Constructor

        :param N: sliding window width.
                  The algorithm will return an estimate of the number of "True"
                  in the last N elements of the stream.
        :type N: int
        :param error_rate: the maximum error made by the algorithm.
        The error rate is in ]0, 1]
        Let c be the true result and e the estimate returned by the dgim
        algorithm.
        abs(c-e) < error_rate * c
        The lower the error, the higher the object footprint.
        :type error_rate: float
        """
        self.N = N

        if not (0 < error_rate <= 1):
            error_msg = ("Invalid value for error_rate: {}. "
                         "Error rate should be in ]0, 1].".format(error_rate))
            raise ValueError(error_msg)

        self.error_rate = error_rate

        #the maximum number of buckets of the same size
        self._r = math.ceil(1/error_rate)
        self._r = max(self._r, 2)

        # the datastructure to store the buckets
        # it is an array of queues
        # queue[i] stores the timestamp of the bucket of size 2^i
        # in descending order
        # this structure makes it easy to :
        # - know how many bucket of the same size there is
        # - update a queue of bucket
        self._queues = []
        if N == 0:
            max_index = -1
        else:
            max_index = int(math.ceil(math.log(N)/math.log(2)))

        self._queues = [deque() for _ in range(max_index + 1)]

        self._timestamp = 0
        self._oldest_bucket_timestamp = -1  # No bucket so far

    def update(self, elt):
        """Update the stream with one element.

        :param elt: the latest element of the stream
        :type elt: bool
        """
        if self.N == 0:
            return
        self._timestamp = (self._timestamp + 1) % (2 * self.N)
        #check if oldest bucket should be removed
        if (self._oldest_bucket_timestamp >= 0 and
                self._is_bucket_too_old(self._oldest_bucket_timestamp)):
            self._drop_oldest_bucket()
        if elt is not True:
            #nothing to do
            return

        carry_over = self._timestamp
        if self._oldest_bucket_timestamp == -1:
            self._oldest_bucket_timestamp = self._timestamp
        for queue in self._queues:
            queue.appendleft(carry_over)
            if len(queue) <= self._r:
                break
            last = queue.pop()
            second_last = queue.pop()
            # merge last two buckets.
            carry_over = second_last
            if last == self._oldest_bucket_timestamp:
                self._oldest_bucket_timestamp = second_last

    def get_count(self):
        """Returns an estimate of the number of "True"
        in the last N elements of the stream.

        :rtype: int
        """
        result = 0
        max_value = 0
        power_of_two = 1
        for queue in self._queues:
            queue_length = len(queue)
            if queue_length > 0:
                max_value = power_of_two
                result += queue_length * power_of_two
            power_of_two = power_of_two << 1
        result -= math.floor(max_value/2)
        return int(result)

    def _is_bucket_too_old(self, bucket_timestamp):
        """Check if a bucket is too old and should be dropped.

        ;param bucket_timestamp: the bucket timestamp
        :type bucket_timestamp: int
        :rtype: bool
        """
        # the buckets are stored modulo 2 * N
        return (self._timestamp - bucket_timestamp) % (2 * self.N) >= self.N

    @property
    def nb_buckets(self):
        """Returns the number of buckets.

        :rtype: int
        """
        result = 0
        for queue in self._queues:
            result += len(queue)
        return result

    def _drop_oldest_bucket(self):
        """Drop oldest bucket timestamp."""
        for queue in reversed(self._queues):
            if len(queue) > 0:
                queue.pop()
                break
        #update oldest bucket timestamp
        self._oldest_bucket_timestamp = -1
        for queue in reversed(self._queues):
            if len(queue) > 0:
                self._oldest_bucket_timestamp = queue[-1]
                break


